File: go/cote/GJM/외발뛰기.go
package main

import (
	"fmt"
	"math/rand"
)

const (
	TEST_CASES = 10
	N = 100
	MAX_VALUE = 10
)

var cache [][]int

func initialize(n int, random *rand.Rand) [][]int {

	ret := make([][]int, n)

	for i := 0; i < n; i++ {
		ret[i] = make([]int, n)

		for j := 0; j < n; j++ {
			ret[i][j] = random.Intn(MAX_VALUE - 1) + 1
		}
	}

	return ret
}

func search(map_ [][]int, y, x int) bool {
	n := len(map_)
	
	if y >= n || x >= n {
		return false
	}

	if y == n - 1 && x == n - 1 {
		return true
	}

	ret := cache[y][x]

	if ret != -1 {
		return ret == 1
	}

	jump := map_[y][x]
	result := search(map_, y + jump, x) || search(map_, y, x + jump)

	if result {
		cache[y][x] = 1
	} else {
		cache[y][x] = 0
	}

	return result
}

func solve(map_ [][]int) bool {
	cache = make([][]int, len(map_))
	
	for i, _ := range map_ {
		cache[i] = make([]int, len(map_))
		for j, _ := range cache[i] {
			cache[i][j] = -1
		}
	}

	return search(map_, 0, 0)
}

func main() {
	random := rand.New(rand.NewSource(43))

	for it := 0; it < TEST_CASES; it++ {
		n := random.Intn(N - 1) + 1
		map_ := initialize(n, random)

		fmt.Println(solve(map_))
	}
}	

File: c++/cote/GJM/외발뛰기.cpp
#include <iostream>
#include <random>
#include <vector>
#define N 100
#define TEST_CASES 10

using namespace std;

int cache[N][N];

int** initialize(default_random_engine& gen) {
  uniform_int_distribution<int> dis(11, 23);

  int** map = new int*[N];

  for (int i = 0; i < N; i++) {
    map[i] = new int[N];
  }

  for (int i = 0; i < N; i++) {
    for (int j = 0; j < N; j++) {
      map[i][j] = dis(gen);
    }
  }

  return map;
}

bool solve(int** map, int x, int y) {
  if (x >= N || y >= N) return false;
  if (x == N - 1 && y == N - 1) return true;

  int jump = map[y][x];

  return (solve(map, x + jump, y) || solve(map, x, y + jump));
}

int solve2(int** map, int x, int y) {
  if (x >= N || y >= N) return 0;
  if (x == N - 1 && y == N - 1) return 1;

  int& ret = cache[y][x];
  if (ret != -1) return ret;
  int jump = map[y][x];

  return (ret = solve2(map, x + jump, y) || solve2(map, x, y + jump));
}

int main(void) {
  default_random_engine gen(43);

  for (int i = 0; i < TEST_CASES; i++) {
    int** map = initialize(gen);
    memset(cache, -1, sizeof(cache));

    cout << boolalpha << solve(map, 0, 0) << "," << solve2(map, 0, 0) << endl;

    for (int i = 0; i < N; i++) {
      delete[] map[i];
    }

    delete[] map;
  }

  return 0;
}